app/src/main/java/com/sschuraytz/charcoaldrawing/UndoRedo.java

```diff
@@ -0,0 +1,63 @@
+package com.sschuraytz.charcoaldrawing;
+
+import android.graphics.Bitmap;
+import android.graphics.Canvas;
+
+import java.util.Stack;
+
+public class UndoRedo {
+
+    private Stack<Bitmap> currentStack = new Stack<>();
+    private Stack<Bitmap> undoneStack = new Stack<>();
+    private Canvas bitmapCanvas;
+
+    public UndoRedo() {
+        onSizeChanged(200, 200);
+    }
+
+    //ensure we always have a bitmap to display
+    protected void onSizeChanged(int width, int height) {
+        currentStack.clear();
```

Comment:
Clear the undone stack as well

Reply:
> Clear the undone stack as well

Even though I clear undoneStack in addBitmap() which gets called in every ACTION_DOWN?

Reply:
OnSizeChanged() can get called I'd you rotate the device (with certain settings).
If you are changing the size of the bitmaps  it's a good idea to clear out all bitmaps you have saved anywhere.